performance-tuning-practices.md

```diff
@@ -0,0 +1,439 @@
+
+- Database Time by SQL Type: the `Select` statement type takes the most time, followed by `general` statements.
+- Database Time by SQL Phase: the `execute` and `compile` phases take most of the time.
+- SQL Execute Time Overview: `Get`, `Cop`, `Prewrite`, and `tso wait` take most of the time.
+- CPS By Type: 3 types of commands (`StmtPrepare`, `StmtExecute`, `StmtClose`) are used.
+- avg QPS = 19.7k (from 24.4k to 19.7k)
+- The execution plan cache is not hit.
+
+The TiDB average CPU utilization increases from 874% to 936%.
+
+![performance-overview-1-for-PrepStmts](/media/performance/3-2.png)
+
+The key latency metrics are as follows:
+
+![performance-overview-2-for-PrepStmts](/media/performance/3.4.png)
+
+- avg query duration = 528μs (from 1.12ms to 528μs)
+- avg parse duration = 14.9μs (from 84.7μs to 14.9μs)
+- avg compile duration = 374μs (from 370μs to 374μs)
+- avg execution duration = 649μs (from 626μs to 649μs)
+
+### Analysis conclusion
+
+Unlike Scenario 2, the application in Scenario 3 enables the Prepared Statement interface but still fails to hit the cache. In addition, Scenario 2 has only one CPS By Type command type (`Query`), while Scenario 3 has three more command types (`StmtPrepare`, `StmtExecute`, `StmtClose`). Compared with Scenario 2, Scenario 3 has two more network round-trip delays.
+
+- Analysis for the decrease in QPS: From the **CPS By Type** pane, you can see that Scenario 2 has only one CPS By Type command type (`Query`), while Scenario 3 has three more command types (`StmtPrepare`, `StmtExecute`, `StmtClose`). `StmtPrepare` and `StmtClose` are non-conventional commands that are not counted by QPS, so QPS is reduced. The non-conventional commands `StmtPrepare` and `StmtClose` are counted in the `general` SQL type, so `general` time is displayed in the database overview of Scenario 3, and it accounts for more than a quarter of the database time.
+- Analysis for the significant decrease in average query duration: for the `StmtPrepare` and `StmtClose` command types newly added in Scenario 3, their query duration is calculated separately in the TiDB internal processing. TiDB executes these two types of commands very quickly, so the average query duration is significantly reduced.
+
+Although Scenario 3 uses the Prepare Statement interface, the execution plan cache is still not hit, because many application frameworks call the `StmtClose` method after `StmtExecute` to prevent memory leaks. Starting from v6.0.0, you can set the global variable `tidb_ignore_prepared_cache_close_stmt=on;`. After that, TiDB will not clear the cached execution plans even if the application calls the `StmtClose` method, so the next SQL execution can reuse an existing execution plan and avoid compiling the execution plan repeatedly.
```

Comment:
```suggestion
Although Scenario 3 uses the Prepared Statement interface, the execution plan cache is still not hit, because many application frameworks call the `StmtClose` method after `StmtExecute` to prevent memory leaks. Starting from v6.0.0, you can set the global variable `tidb_ignore_prepared_cache_close_stmt=on;`. After that, TiDB will not clear the cached execution plans even if the application calls the `StmtClose` method, so the next SQL execution can reuse the existing execution plan and avoid compiling the execution plan repeatedly.
```